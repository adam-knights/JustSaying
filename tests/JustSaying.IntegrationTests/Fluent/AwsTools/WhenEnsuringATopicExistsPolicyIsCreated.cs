using Amazon.SQS.Model;
using JustSaying.AwsTools;
using JustSaying.AwsTools.QueueCreation;
using JustSaying.TestingFramework;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json.Linq;

#pragma warning disable 618

namespace JustSaying.IntegrationTests.Fluent.AwsTools;

public class WhenEnsuringATopicExistsPolicyIsCreated : IntegrationTestBase
{
    public WhenEnsuringATopicExistsPolicyIsCreated(ITestOutputHelper outputHelper)
        : base(outputHelper)
    {
    }

    [AwsFact]
    public async Task Then_The_Iam_Policy_Is_Created()
    {
        // Arrange
        ILoggerFactory loggerFactory = OutputHelper.ToLoggerFactory();
        IAwsClientFactory clientFactory = CreateClientFactory();
        var QueueName = UniqueName;
        var TopicName = $"{UniqueName}-topic";

        var client = clientFactory.GetSqsClient(Region);

        var queueCreator = new AmazonQueueCreator(new AwsClientFactoryProxy(() => clientFactory), loggerFactory);

        var readConfig = new SqsReadConfiguration(SubscriptionType.ToTopic)
        {
            QueueName = QueueName,
            SubscriptionGroupName = QueueName,
            TopicName = TopicName,
            PublishEndpoint = TopicName,
            RetryCountBeforeSendingToErrorQueue = 1,
        };

        var queue = queueCreator.EnsureTopicExistsWithQueueSubscribed(RegionName, readConfig);

        // Act
        await queue.StartupTask.Invoke(CancellationToken.None);

        // Assert
        var queueAttributesResponse = await client.GetQueueAttributesAsync(new GetQueueAttributesRequest
        {
            QueueUrl = queue.Queue.Uri.ToString(),
            AttributeNames = new List<string> { "Policy" }
        });
        queueAttributesResponse.Policy.ShouldMatchApproved(opt =>
        {
            opt.SubFolder("Approvals");

            opt.WithScrubber(policy => ScrubUniqueData(policy, UniqueName, RegionName));
        });
    }

    private static string ScrubUniqueData(string iamPolicy, string uniqueName, string regionName)
    {
        var json = JObject.Parse(iamPolicy);
        var sourceArn = json["Statement"]![0]!["Condition"]!["ArnLike"]!["aws:SourceArn"]!.ToString();
        var resourceArn = json["Statement"]![0]!["Resource"]!.ToString();
        return iamPolicy
            .Replace(json["Statement"]![0]!["Sid"]!.ToString(), "<sid>")
            .Replace(resourceArn, resourceArn.Replace(uniqueName, "<unique-name>"))
            .Replace(sourceArn, sourceArn.Replace(uniqueName, "<unique-name>"));
    }
}
