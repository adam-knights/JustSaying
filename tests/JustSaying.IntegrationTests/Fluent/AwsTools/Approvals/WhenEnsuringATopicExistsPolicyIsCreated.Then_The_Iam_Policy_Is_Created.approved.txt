{
    "Version" : "2012-10-17",
    "Statement" : [
        {
            "Sid" : "<sid>",
            "Effect" : "Allow",
            "Principal" : {
                "AWS" : "*"
            },
            "Action"    : "sqs:SendMessage",
            "Resource"  : "arn:aws:sqs:us-east-1:000000000000:<unique-name>",
            "Condition" : {
                "ArnLike" : {
                    "aws:SourceArn" : "arn:aws:sns:us-east-1:000000000000:<unique-name>-topic"
                }
            }
        }
    ]
}